File: macOS_app/Sources/PSoCDriverWrapper/PSoCDriver.swift
import Foundation
import SwiftUI

// Swift wrapper for PSoC Driver C library

public struct PSoCProtocol {
    public static let serviceUUID = "A1B2C3D4-E5F6-4A5B-8C9D-0E1F2A3B4C5D"
    public static let dataBlockUUID = "A1B2C3D5-E5F6-4A5B-8C9D-0E1F2A3B4C5D"
    public static let controlUUID = "A1B2C3D6-E5F6-4A5B-8C9D-0E1F2A3B4C5D"
    public static let deviceName = "Inductosense Temp"
    public static let totalBlocks: Int32 = 1800
    public static let blockSize: Int32 = 7168
    public static let ackInterval: Int32 = 20
}

public struct PSoCWaveformHeader {
    public let blockNumber: UInt32
    public let timestampMs: UInt32
    public let sampleRateHz: UInt32
    public let sampleCount: UInt16
    public let triggerSample: UInt16
    public let pulseFreqHz: UInt32
    public let temperatureCx10: Int16
    public let gainDb: UInt8

    init(from cHeader: waveform_header_t) {
        self.blockNumber = cHeader.block_number
        self.timestampMs = cHeader.timestamp_ms
        self.sampleRateHz = cHeader.sample_rate_hz
        self.sampleCount = cHeader.sample_count
        self.triggerSample = cHeader.trigger_sample
        self.pulseFreqHz = cHeader.pulse_freq_hz
        self.temperatureCx10 = cHeader.temperature_cx10
        self.gainDb = cHeader.gain_db
    }
}

public struct PSoCWaveform {
    public let header: PSoCWaveformHeader
    public let samples: [Int32]
    public let isCompressed: Bool

    init(from cWaveform: waveform_data_t, isCompressed: Bool) {
        self.header = PSoCWaveformHeader(from: cWaveform.header)
        var mutableWaveform = cWaveform
        self.samples = withUnsafeBytes(of: &mutableWaveform.samples) { bytes in
            let buffer = bytes.bindMemory(to: Int32.self)
            return Array(buffer)
        }
        self.isCompressed = isCompressed
    }
}

public struct PSoCTransferStats {
    public let blocksReceived: UInt32
    public let totalBlocks: UInt32
    public let totalBytesReceived: UInt32
    public let totalChunksReceived: UInt32
    public let throughputKBps: Double
    public let progressPercent: Double
    public let elapsedSeconds: Double

    init(from cStats: transfer_stats_t) {
        self.blocksReceived = cStats.blocks_received
        self.totalBlocks = cStats.total_blocks
        self.totalBytesReceived = cStats.total_bytes_received
        self.totalChunksReceived = cStats.total_chunks_received
        self.throughputKBps = cStats.throughput_kbps
        self.progressPercent = cStats.progress_percent
        self.elapsedSeconds = cStats.elapsed_seconds
    }
}

public class PSoCTransferSession {
    private var session: OpaquePointer?
    public var onWaveform: ((PSoCWaveform) -> Void)?
    public var onProgress: ((PSoCTransferStats) -> Void)?
    public var onCompletion: ((PSoCTransferStats) -> Void)?
    public var onAck: ((UInt16) -> Void)?

    public init() {
        session = transfer_session_create()
        setupCallbacks()
    }

    deinit {
        if let session = session {
            transfer_session_destroy(session)
        }
    }

    private func setupCallbacks() {
        guard let session = session else { return }

        // Waveform callback
        let waveformContext = Unmanaged.passUnretained(self).toOpaque()
        transfer_session_set_waveform_callback(session, { waveformPtr, isCompressed, userData in
            guard let waveformPtr = waveformPtr, let userData = userData else { return }
            let selfRef = Unmanaged<PSoCTransferSession>.fromOpaque(userData).takeUnretainedValue()
            let waveform = waveformPtr.pointee
            let swiftWaveform = PSoCWaveform(from: waveform, isCompressed: isCompressed)
            DispatchQueue.main.async {
                selfRef.onWaveform?(swiftWaveform)
            }
        }, waveformContext)

        // Progress callback
        let progressContext = Unmanaged.passUnretained(self).toOpaque()
        transfer_session_set_progress_callback(session, { statsPtr, userData in
            guard let statsPtr = statsPtr, let userData = userData else { return }
            let selfRef = Unmanaged<PSoCTransferSession>.fromOpaque(userData).takeUnretainedValue()
            let stats = PSoCTransferStats(from: statsPtr.pointee)
            DispatchQueue.main.async {
                selfRef.onProgress?(stats)
            }
        }, progressContext)

        // Completion callback
        let completionContext = Unmanaged.passUnretained(self).toOpaque()
        transfer_session_set_completion_callback(session, { statsPtr, userData in
            guard let statsPtr = statsPtr, let userData = userData else { return }
            let selfRef = Unmanaged<PSoCTransferSession>.fromOpaque(userData).takeUnretainedValue()
            let stats = PSoCTransferStats(from: statsPtr.pointee)
            DispatchQueue.main.async {
                selfRef.onCompletion?(stats)
            }
        }, completionContext)

        // ACK callback
        let ackContext = Unmanaged.passUnretained(self).toOpaque()
        transfer_session_set_ack_callback(session, { blockNumber, userData in
            guard let userData = userData else { return }
            let selfRef = Unmanaged<PSoCTransferSession>.fromOpaque(userData).takeUnretainedValue()
            selfRef.onAck?(blockNumber)
        }, ackContext)
    }

    public func start() {
        guard let session = session else { return }
        transfer_session_start(session)
    }

    public func stop() {
        guard let session = session else { return }
        transfer_session_stop(session)
    }

    public func processChunk(data: Data) -> Bool {
        guard let session = session else { return false }
        return data.withUnsafeBytes { bytes in
            guard let baseAddress = bytes.baseAddress else { return false }
            return transfer_session_process_chunk(session, baseAddress.assumingMemoryBound(to: UInt8.self), bytes.count)
        }
    }

    public func getStats() -> PSoCTransferStats? {
        guard let session = session else { return nil }
        var stats = transfer_stats_t()
        transfer_session_get_stats(session, &stats)
        return PSoCTransferStats(from: stats)
    }

    public var isActive: Bool {
        guard let session = session else { return false }
        return transfer_session_is_active(session)
    }
}

File: Windows_app/Interop/PSoCDriver.cs
using System;
using System.Runtime.InteropServices;
using System.Windows;

namespace BLETester.Interop
{
    /// <summary>
    /// Managed wrapper for waveform header
    /// </summary>
    public class WaveformHeader
    {
        public uint BlockNumber { get; set; }
        public uint TimestampMs { get; set; }
        public uint SampleRateHz { get; set; }
        public ushort SampleCount { get; set; }
        public ushort TriggerSample { get; set; }
        public uint PulseFreqHz { get; set; }
        public short TemperatureCx10 { get; set; }
        public byte GainDb { get; set; }

        internal static WaveformHeader FromNative(NativeMethods.WaveformHeader native)
        {
            return new WaveformHeader
            {
                BlockNumber = native.BlockNumber,
                TimestampMs = native.TimestampMs,
                SampleRateHz = native.SampleRateHz,
                SampleCount = native.SampleCount,
                TriggerSample = native.TriggerSample,
                PulseFreqHz = native.PulseFreqHz,
                TemperatureCx10 = native.TemperatureCx10,
                GainDb = native.GainDb
            };
        }
    }

    /// <summary>
    /// Managed wrapper for waveform data
    /// </summary>
    public class Waveform
    {
        public WaveformHeader Header { get; set; }
        public int[] Samples { get; set; }
        public bool IsCompressed { get; set; }

        public Waveform(WaveformHeader header, int[] samples, bool isCompressed)
        {
            Header = header;
            Samples = samples;
            IsCompressed = isCompressed;
        }
    }

    /// <summary>
    /// Managed wrapper for transfer statistics
    /// </summary>
    public class TransferStats
    {
        public uint BlocksReceived { get; set; }
        public uint TotalBlocks { get; set; }
        public uint TotalBytesReceived { get; set; }
        public uint TotalChunksReceived { get; set; }
        public double ThroughputKBps { get; set; }
        public double ProgressPercent { get; set; }
        public double ElapsedSeconds { get; set; }

        internal static TransferStats FromNative(NativeMethods.TransferStats native)
        {
            return new TransferStats
            {
                BlocksReceived = native.BlocksReceived,
                TotalBlocks = native.TotalBlocks,
                TotalBytesReceived = native.TotalBytesReceived,
                TotalChunksReceived = native.TotalChunksReceived,
                ThroughputKBps = native.ThroughputKBps,
                ProgressPercent = native.ProgressPercent,
                ElapsedSeconds = native.ElapsedSeconds
            };
        }
    }

    /// <summary>
    /// Managed wrapper for PSoC transfer session
    /// </summary>
    public class TransferSession : IDisposable
    {
        private IntPtr _session;
        private NativeMethods.WaveformCallback _waveformCallback;
        private NativeMethods.ProgressCallback _progressCallback;
        private NativeMethods.CompletionCallback _completionCallback;
        private NativeMethods.AckCallback _ackCallback;

        public event Action<Waveform> OnWaveform;
        public event Action<TransferStats> OnProgress;
        public event Action<TransferStats> OnCompletion;
        public event Action<ushort> OnAck;

        public TransferSession()
        {
            _session = NativeMethods.transfer_session_create();
            if (_session == IntPtr.Zero)
                throw new InvalidOperationException("Failed to create transfer session");

            SetupCallbacks();
        }

        private void SetupCallbacks()
        {
            // Store delegates as fields to prevent garbage collection
            _waveformCallback = (waveformPtr, isCompressed, userData) =>
            {
                try
                {
                    if (waveformPtr == IntPtr.Zero) return;

                    // Read header
                    var header = Marshal.PtrToStructure<NativeMethods.WaveformHeader>(waveformPtr);
                    var managedHeader = WaveformHeader.FromNative(header);

                    // Read samples (2376 int32 values after header)
                    var samplesPtr = IntPtr.Add(waveformPtr, Marshal.SizeOf<NativeMethods.WaveformHeader>());
                    var samples = new int[2376];
                    Marshal.Copy(samplesPtr, samples, 0, 2376);

                    var waveform = new Waveform(managedHeader, samples, isCompressed);

                    // Invoke on UI thread
                    Application.Current?.Dispatcher.Invoke(() => OnWaveform?.Invoke(waveform));
                }
                catch (Exception ex)
                {
                    System.Diagnostics.Debug.WriteLine($"Error in waveform callback: {ex.Message}");
                }
            };

            _progressCallback = (statsPtr, userData) =>
            {
                try
                {
                    if (statsPtr == IntPtr.Zero) return;

                    var stats = Marshal.PtrToStructure<NativeMethods.TransferStats>(statsPtr);
                    var managedStats = TransferStats.FromNative(stats);

                    Application.Current?.Dispatcher.Invoke(() => OnProgress?.Invoke(managedStats));
                }
                catch (Exception ex)
                {
                    System.Diagnostics.Debug.WriteLine($"Error in progress callback: {ex.Message}");
                }
            };

            _completionCallback = (statsPtr, userData) =>
            {
                try
                {
                    if (statsPtr == IntPtr.Zero) return;

                    var stats = Marshal.PtrToStructure<NativeMethods.TransferStats>(statsPtr);
                    var managedStats = TransferStats.FromNative(stats);

                    Application.Current?.Dispatcher.Invoke(() => OnCompletion?.Invoke(managedStats));
                }
                catch (Exception ex)
                {
                    System.Diagnostics.Debug.WriteLine($"Error in completion callback: {ex.Message}");
                }
            };

            _ackCallback = (blockNumber, userData) =>
            {
                OnAck?.Invoke(blockNumber);
            };

            // Register callbacks
            NativeMethods.transfer_session_set_waveform_callback(_session, _waveformCallback, IntPtr.Zero);
            NativeMethods.transfer_session_set_progress_callback(_session, _progressCallback, IntPtr.Zero);
            NativeMethods.transfer_session_set_completion_callback(_session, _completionCallback, IntPtr.Zero);
            NativeMethods.transfer_session_set_ack_callback(_session, _ackCallback, IntPtr.Zero);
        }

        public void Start()
        {
            NativeMethods.transfer_session_start(_session);
        }

        public void Stop()
        {
            NativeMethods.transfer_session_stop(_session);
        }

        public bool ProcessChunk(byte[] data)
        {
            return NativeMethods.transfer_session_process_chunk(_session, data, (UIntPtr)data.Length);
        }

        public TransferStats GetStats()
        {
            NativeMethods.transfer_session_get_stats(_session, out var stats);
            return TransferStats.FromNative(stats);
        }

        public bool IsActive
        {
            get { return NativeMethods.transfer_session_is_active(_session); }
        }

        public void Dispose()
        {
            if (_session != IntPtr.Zero)
            {
                NativeMethods.transfer_session_destroy(_session);
                _session = IntPtr.Zero;
            }
            GC.SuppressFinalize(this);
        }

        ~TransferSession()
        {
            Dispose();
        }
    }
}
